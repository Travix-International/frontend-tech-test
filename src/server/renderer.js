import React from 'react';
import ReactServer from 'react-dom/server';
import { StaticRouter as Router, matchPath } from 'react-router';
import { Provider } from 'react-redux';
import { Helmet } from 'react-helmet';
import { /* get as _get, */ isFunction as _isFunction } from 'lodash';
import renderFullPage from './layout';
import configureStore from '../configureStore';
import { entryRouteComponentMap, getEntryAndRoute, entryReducerMap } from './entryAndRoute';

function applyInitStore(req) {
  const entryRouteInfo = getEntryAndRoute(req.path);
  const currentEntry = entryRouteInfo.entry;
  const rootReducer = entryReducerMap[currentEntry];
  const store = configureStore({}, rootReducer);
  req.reduxStore = store;
  req.entryRouteInfo = entryRouteInfo;
  // next();
}

function applyRouteCheckResult(req) {
  const { path, reduxStore: store, query, entryRouteInfo } = req;
  const promises = [];
  const currentEntry = entryRouteInfo.entry;
  const route = entryRouteInfo.route;
  const match = matchPath(path, route);
  const { loadData } = route;

  // set currentEntry to req
  req.currentEntry = currentEntry;

  // it's possible no loadData set
  if (_isFunction(loadData)) {
    let action = loadData(match, query);
    if (action) {
      if (_isFunction(action) || action.type) {
        action = store.dispatch(action);
        if (action.then) {
          promises.push(action);
        }
      }
    }
  }

  return Promise.all(promises);
}

function responsePage(req, res, clientStats) {
  const { url, reduxStore: store, currentEntry } = req;
  const routerContext = {};
  const reduxStateString = JSON.stringify(store.getState()).replace(/</g, '\\u003c');

  const RouteComponent = entryRouteComponentMap[currentEntry];

  const content = (
    <Provider store={store}>
      <Router context={routerContext} location={url}>
        <RouteComponent />
      </Router>
    </Provider>
  );

  // render to sting to get helmet setting
  const contentString = ReactServer.renderToString(content);
  const head = Helmet.renderStatic();
  const htmlString = renderFullPage(contentString, reduxStateString, head, currentEntry, clientStats);

  res.send(htmlString);
}

export default function renderer({ clientStats }) {
  return (req, res) => {
    applyInitStore(req);
    return applyRouteCheckResult(req).then(() => {
      return responsePage(req, res, clientStats);
    });
  };
}
