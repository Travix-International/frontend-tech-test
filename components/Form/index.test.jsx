import React from 'react';
import renderer from 'react-test-renderer';
import { FormRender } from './index';

describe('Form snapshot test', () => {
  it('DOM structure renders correctly', () => {
    const mockFn = jest.fn();
    const tree = renderer
      .create(<FormRender
        action={mockFn}
        actionBtnTitle="Action"
        changeDescription={mockFn}
        changeTitle={mockFn}
        clearDescription={mockFn}
        clearTitle={mockFn}
        description="Description"
        formTitle="Form Title"
        title="Title"
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
