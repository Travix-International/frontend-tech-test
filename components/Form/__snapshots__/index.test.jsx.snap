// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshot test DOM structure renders correctly 1`] = `
<form
  className="ui-form"
  onSubmit={[Function]}
>
  <label
    htmlFor="titleInput"
  >
    Form Title
  </label>
  <div
    className="ui-input-container"
  >
    <input
      className="ui-input"
      disabled={false}
      id="titleInput"
      name={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Todo title..."
      type="text"
      value="Title"
    />
  </div>
  <div
    className="ui-input-container"
  >
    <textarea
      className="ui-input"
      disabled={false}
      name={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Todo description..."
      type="text"
      value="Description"
    />
  </div>
  <button
    className="ui-button ui-button_size_s ui-button_variation_default"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Action
  </button>
</form>
`;
